node/core/approval-voting/src/lib.rs

```diff
@@ -539,6 +534,117 @@ struct ApprovalStatus {
 	block_tick: Tick,
 }
 
+#[derive(Copy, Clone)]
+enum ApprovalOutcome {
+	Approved,
+	Failed,
+	TimedOut,
+}
+
+struct ApprovalState {
+	validator_index: ValidatorIndex,
+	candidate_hash: CandidateHash,
+	candidate_index: CandidateIndex,
+	approval_outcome: ApprovalOutcome,
+}
+
+impl ApprovalState {
+	fn approved(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex,
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			approval_outcome: ApprovalOutcome::Approved,
+		}
+	}
+	fn failed(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex,
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			approval_outcome: ApprovalOutcome::Failed,
+		}
+	}
+}
+
+struct CurrentlyCheckingSet {
+	candidate_hash_map: HashMap<CandidateHash, Vec<Hash>>,
+	currently_checking: FuturesUnordered<BoxFuture<'static, ApprovalState>>,
+}
+
+impl Default for CurrentlyCheckingSet {
+	fn default() -> Self {
+		Self {
+			candidate_hash_map: HashMap::new(),
+			currently_checking: FuturesUnordered::new(),
+		}
+	}
+}
+
+impl CurrentlyCheckingSet {
+	// The return value of this function will indicate whether we have seen this
+	// candidate_hash for this relay_block before. If we have never seen this candidate_hash
+	// xor never seen this relay_block (for this candidate_hash), the function will return false
+	pub async fn insert_relay_block_hash(
+		&mut self,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex,
+		validator_index: ValidatorIndex,
+		relay_block: Hash,
+		launch_work: impl Future<Output = SubsystemResult<RemoteHandle<ApprovalState>>>,
+	) -> SubsystemResult<()> {
+		let val = self.candidate_hash_map
+			.entry(candidate_hash)
+			.or_insert(Default::default());
+
+		if let Err(k) = val.binary_search_by_key(&relay_block, |v| *v) {
+			let _ = val.insert(k, relay_block);
+			let work = launch_work.await?;
+			self.currently_checking.push(
+				Box::pin(async move {
+					match work.timeout(Duration::new(APPROVAL_CHECKING_TIMEOUT, 0)).await {
+						None => ApprovalState {
+							candidate_hash,
+							validator_index,
+							candidate_index,
```

Comment:
`candidate_index` being in ApprovalState is a bug - we're going to reuse the same ApprovalState for multiple blocks, which contain the same candidate but at different indices. 

What is ApprovalState.candidate_index used for? Maybe it can be elided?